### ISSUE TITLE
Message Utilities

### ISSUE THREAD

**Issue by IGJoshua** [IGJoshua/discljord#42]
Currently things like mentions, channel mentions, custom emoji, and other text formatting for messages is assumed that the user will perform it on their own.
Discljord can and should provide utility functions for constructing message bodies which have correct formatting for these types of use cases, allowing the user to know what they want to do without having to browse the discord documentation to learn how to do it.

**Comment by JohnnyJayJay** [comment]
What do you have in mind in terms of design? Looking at mention formatting, I've been thinking of a variety of functions such as `mention-user`, `mention-emote` and so on that can either take an entire entity map or just an id. For further formatting, there could be markdown tools (that sanitise/escape all markdown for example), a function to get `username#discriminator` from a user object etc. 

Additionally, something I've found myself needing often is the avatar/icon url of a user/guild, which you currently still have to make yourself. That could maybe fit into this issue, although it's not really part of message formatting.

**Comment by IGJoshua** [comment]
So I think that depends somewhat on the particular function. For something like a user or channel mention, it makes the most sense to just take the id, since it only ever works in one way and it takes one piece of data. For emotes specifically though I think there's more room for discussion. Personally I'm inclined to have it take an emote object, a map with a name and an optional id, that way whatever you save from reaction add or whatever will be able to be used directly, regardless of what type of emote it is, normal vs custom.

As for those additional functions for markdown, `username#discriminator`, etc. I'm absolutely on-board with adding those, and that's part of what I wanted for this issue.

As for the avatar/icon url, I think that those are good to include, although I wouldn't consider it to be a part of this issue in particular, although I'd more than likely accept them as a part of a PR which would close out this issue.

**Comment by JohnnyJayJay** [comment]
> For something like a user or channel mention, it makes the most sense to just take the id, since it only ever works in one way and it takes one piece of data.

Yes, that's true, but it might be more convenient to just be able to pass it the entity directly in cases where you have it already instead of picking out the id. The functions for those kinds of metions could look something like this maybe:
```clj
(defn mention-user [user]
  (str "<@" (cond-> user (map? user) :id) \>))
```
But I understand if you want to keep it simple and only accept ids.

> For emotes specifically though I think there's more room for discussion. Personally I'm inclined to have it take an emote object, a map with a name and an optional id, that way whatever you save from reaction add or whatever will be able to be used directly, regardless of what type of emote it is, normal vs custom.

Since you only need an id to mention custom emojis (non-animated ones, at least), it would be nice to have support for just an id like in the other cases, I think. As for map arguments, making it so that it checks whether an id is present and based on that either returning the name directly or constructing a mention sounds like a good idea.

**Comment by IGJoshua** [comment]
How do you mean you only need an id to mention custom emojis? All non-animated custom emoji mentions are of the form `<:name:some-id>` and lacking a name doesn't produce an emoji when sent. If they are animated, then it's of the form `<a:name:some-id>`. Based on that, I think that the emoji one should always take only a map. As for the other functions, I don't mind adding the check if they are a map which contains the `:id` key and fetching that if found. That way it can be either the id or the object, and the user shouldn't have to care which.

https://discord.com/developers/docs/reference#message-formatting

**Comment by JohnnyJayJay** [comment]
You need to put something in for the name, yes, but it can be anything. It will not be taken into account. I.e., `<:_:id>` works for any custom emoji (except animated ones because the `a` prefix is missing). I [leveraged this in my poker bot](https://github.com/JohnnyJayJay/poker/blob/master/src/poker/discord/display.clj#L59) for example.

> As for the other functions, I don't mind adding the check if they are a map which contains the :id key and fetching that if found. That way it can be either the id or the object, and the user shouldn't have to care which.

Yeah, that's what I was going for.

**Comment by IGJoshua** [comment]
Oh that's *interesting*. I wonder what bots will see if you do that. In any case, I approve of that taking an id as well then.

**Comment by JohnnyJayJay** [comment]
Discord inserts the correct name before sending it to other users.

**Comment by IGJoshua** [comment]
This has been implemented with #43
